packages/shared/scope-hoisting/src/link.js

```diff
@@ -140,233 +150,369 @@ export function link({
       }
     }
 
+    for (let [symbol, {local}] of asset.symbols) {
+      exports.set(local, [asset, symbol]);
+    }
+
     if (bundleGraph.isAssetReferencedByDependant(bundle, asset)) {
       referencedAssets.add(asset);
     }
   });
 
+  let entry = bundle.getMainEntry();
+  let exportedSymbols: Map<
+    string,
+    Array<{|exportAs: string, local: string|}>,
+  > = new Map();
+  if (entry) {
+    if (entry.meta.isCommonJS) {
+      if (bundle.env.outputFormat === 'commonjs') {
+        exportedSymbols.set(assertString(entry.meta.exportsIdentifier), [
+          {exportAs: '*', local: 'exports'},
+        ]);
+      }
+    } else {
+      for (let {
+        exportAs,
+        exportSymbol,
+        symbol,
+        asset,
+        loc,
+      } of bundleGraph.getExportedSymbols(entry)) {
+        if (typeof symbol === 'string') {
+          let symbols = exportedSymbols.get(
+            symbol === '*'
+              ? assertString(entry.meta.exportsIdentifier)
+              : symbol,
+          );
+
+          let local = exportAs;
+          if (symbols) {
+            local = symbols[0].local;
+          } else {
+            symbols = [];
+            exportedSymbols.set(symbol, symbols);
+
+            if (local === '*') {
+              local = 'exports';
+            } else if (!t.isValidIdentifier(local) || globalNames.has(local)) {
+              local = scope.generateUid(local);
+            } else {
+              scope.add(local);
+            }
+          }
+
+          symbols.push({exportAs, local});
+        } else if (symbol === null) {
+          // TODO `meta.exportsIdentifier[exportSymbol]` should be exported
+          let relativePath = relative(options.projectRoot, asset.filePath);
+          throw getThrowableDiagnosticForNode(
+            `${relativePath} couldn't be statically analyzed when importing '${exportSymbol}'`,
+            entry.filePath,
+            loc,
+          );
+        } else if (symbol !== false) {
+          let relativePath = relative(options.projectRoot, asset.filePath);
+          throw getThrowableDiagnosticForNode(
+            `${relativePath} does not export '${exportSymbol}'`,
+            entry.filePath,
+            loc,
+          );
+        }
+      }
+    }
+  }
+
+  let resolveSymbolCache = new Map();
+
   function resolveSymbol(inputAsset, inputSymbol: Symbol, bundle) {
+    let k = inputAsset.id + ':' + inputSymbol + ':' + bundle.id;
+    let cached = resolveSymbolCache.get(k);
+    if (cached) {
+      return cached;
+    }
+
     let {asset, exportSymbol, symbol, loc} = bundleGraph.resolveSymbol(
       inputAsset,
       inputSymbol,
       bundle,
     );
     if (asset.meta.resolveExportsBailedOut) {
```

Comment:
I think `resolveExportsBailedOut` can be replaced with `pureExports` now. Then `resolveExportsBailedOut` can be removed from meta.